Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.h

```diff
@@ -0,0 +1,1532 @@
+constexpr bool isUnit(U other)
+{
+    if constexpr (std::same_as<T, U>)
+        return true;
+    else if constexpr (NestedUnitEnumOf<U, T>)
+        return true;
+    else if constexpr (NestedUnitEnumOf<T, U>)
+        return UnitTraits<U>::template is<T>(other);
+}
+
+// Allows identity casts, upcasts or downcasts of a unit type.
+template<UnitEnum T, UnitEnum U>
+constexpr T unitCast(U other)
+{
+    if constexpr (std::same_as<T, U>)
+        return other;
+    else if constexpr (NestedUnitEnumOf<U, T>)
+        return UnitTraits<T>::upcast(other);
+    else if constexpr (NestedUnitEnumOf<T, U>) {
+        RELEASE_ASSERT(isUnit<T>(other));
+        return UnitTraits<U>::template downcast<T>(other);
+    }
+}
+
+// Allows unchecked identity casts, upcasts or downcasts of a unit type.
+template<UnitEnum T, UnitEnum U>
+constexpr T uncheckedUnitCast(U other)
+{
+    if constexpr (std::same_as<T, U>)
+        return other;
+    else if constexpr (NestedUnitEnumOf<U, T>)
+        return UnitTraits<T>::upcast(other);
+    else if constexpr (NestedUnitEnumOf<T, U>) {
+        ASSERT(isUnit<T>(other));
+        return UnitTraits<U>::template downcast<T>(other);
+    }
+}
+
+// Allows casting UP from a unit type to a composite unit type that includes it.
+//    `AnglePercentageUnit::Deg` == unitUpcast<AnglePercentageUnit>(AngleUnit::Deg)
+template<UnitEnum T, UnitEnum U>
+constexpr std::optional<T> dynamicUnitCast(U other)
+{
+    if constexpr (std::same_as<T, U>)
+        return other;
+    else if constexpr (NestedUnitEnumOf<U, T>)
+        return UnitTraits<T>::upcast(other);
+    else if constexpr (NestedUnitEnumOf<T, U>) {
+        if (!isUnit<T>(other))
+            return { };
+        return UnitTraits<U>::template downcast<T>(other);
+    }
+}
+
+// Allows casting UP from a unit type to a composite unit type that includes it.
+//    `AnglePercentageUnit::Deg` == unitUpcast<AnglePercentageUnit>(AngleUnit::Deg)
+template<UnitEnum T, UnitEnum U>
+    requires NestedUnitEnumOf<U, T>
+constexpr T unitUpcast(U other)
+{
+    return UnitTraits<T>::upcast(other);
+}
+
+// Allows casting DOWN from a composite unit type to one of the unit types it includes.
+//    `AngleUnit::Deg` == unitDowncast<AngleUnit>(AnglePercentageUnit::Deg)
+//    RELEASE_ASSERT   == unitDowncast<AngleUnit>(AnglePercentageUnit::Percentage)
+// NOTE: This will RELEASE_ASSERT that the cast is valid.
+template<UnitEnum T, UnitEnum U>
+    requires NestedUnitEnumOf<T, U>
+constexpr T unitDowncast(U other)
+{
+    RELEASE_ASSERT(isUnit<T>(other));
+    return UnitTraits<U>::template downcast<T>(other);
+}
+
+// Allows unchecked casting DOWN from a composite unit type to one of the unit types it includes.
+//    `AngleUnit::Deg` == uncheckedUnitDowncast<AngleUnit>(AnglePercentageUnit::Deg)
+//    "undefined"      == uncheckedUnitDowncast<AngleUnit>(AnglePercentageUnit::Percentage)
+// NOTE: Use with cation. Only a debug assertion will check that the cast is valid.
```

Comment:
Offering this goes against trend. We are doing more checking because of the security benefits. This should be here only if there is a really good reason it needs to be.